packages/compiler/emitter-framework/asset-emitter.ts

```diff
@@ -0,0 +1,749 @@
+      );
+
+      const newContext = (typeEmitter as any)[lexicalKey](contextChainEntry);
+      const newReferenceContext = keyHasReferenceContext(key)
+        ? (typeEmitter as any)[referenceKey](contextChainEntry)
+        : {};
+
+      // assemble our new reference and lexical contexts.
+      const newContextState = interner.intern({
+        lexicalContext: interner.intern({
+          ...context.lexicalContext,
+          ...newContext,
+        }),
+        referenceContext: interner.intern({
+          ...context.referenceContext,
+          ...newReferenceContext,
+        }),
+      });
+
+      knownContexts.set([contextChainEntry, context], newContextState);
+      context = newContextState;
+    }
+  }
+
+  /**
+   * Invoke the callback with the proper context for a given type.
+   */
+  function withTypeContext(type: Type, cb: () => void) {
+    const oldContext = context;
+    const oldTypeStack = lexicalTypeStack;
+
+    setContextForType(type);
+    cb();
+
+    context = oldContext;
+    lexicalTypeStack = oldTypeStack;
+  }
+
+  /**
+   * Invoke the callback with the given context.
+   */
+  function withContext(newContext: EmitterState, cb: () => void) {
+    const oldContext = newContext.context;
+    const oldTypeStack = newContext.lexicalTypeStack;
+    context = newContext.context;
+    lexicalTypeStack = newContext.lexicalTypeStack;
+
+    cb();
+
+    context = oldContext;
+    lexicalTypeStack = oldTypeStack;
+  }
+
+  function typeEmitterKey(type: Type) {
+    switch (type.kind) {
+      case "Model":
+        if (type.name === "" || type.name === "Array") {
+          return "modelLiteral";
+        }
+
+        if (type.templateMapper) {
+          return "modelInstantiation";
+        }
+
+        return "modelDeclaration";
+      case "Namespace":
+        return "namespace";
+      case "ModelProperty":
+        return "modelPropertyLiteral";
+      case "Boolean":
+        return "booleanLiteral";
+      case "String":
+        return "stringLiteral";
+      case "Number":
+        return "numericLiteral";
+      case "Operation":
+        if (type.interface) {
+          return "interfaceOperationDeclaration";
+        } else {
+          return "operationDeclaration";
+        }
+      case "Interface":
+        return "interfaceDeclaration";
+      case "Enum":
+        return "enumDeclaration";
+      case "EnumMember":
+        return "enumMember";
+      case "Union":
+        if (!type.name) {
+          return "unionLiteral";
+        }
+
+        if (type.templateMapper) {
+          return "unionInstantiation";
+        }
+
+        return "unionDeclaration";
+      case "UnionVariant":
+        return "unionVariant";
+      case "Tuple":
+        return "tupleLiteral";
+      case "Scalar":
+        return "scalarDeclaration";
+      case "Intrinsic":
+        return "intrinsic";
+      default:
+        compilerAssert(false, `Encountered type ${type.kind} which we don't know how to emit.`);
```

Comment:
worth assert never trick here to catch at compile time?

Reply:
`type.kind` isn't never here, there are some things that shouldn't show up in the type graph that are nonetheless types (`Function`, `TypeParameter`, etc.)